### PR TITLE
feat(orchestrator): Merge local and remote testbed utilities

### LINKED ISSUE #375: Simplify storage node CLI
There is a lot of code repetition between the dry run genesis used to create a local and remote testbed. Also ensure all nodes use the same ports except when running a local testbed (this makes PE live easier)

### PR DESCRIPTION
**Pull request by asonnino**
* Remove code duplication to create genesis
* Book all storage nodes with the same port when deploying remote testbeds

closes #375

### CODE REVIEW

File: Cargo.lock
Comment (mlegner) [inline Cargo.lock:1]: Important: This should not be deleted.

File: crates/walrus-service/bin/main.rs [hunk 1]
```diff
@@ -70,11 +76,11 @@ enum Commands {
...
 
-    /// Generate the configuration files to run a local testbed of storage nodes.
-    GenerateDryRunLocalConfigs(GenerateDryRunLocalConfigsArgs),
+    /// Deploy the Walrus system contract on the Sui network.
+    DeploySystemContract(DeploySystemContractArgs),
 
-    /// Generate the configuration files to run a remote testbed of storage nodes.
-    GenerateDryRunRemoteConfigs(GenerateDryRunRemoteConfigsArgs),
+    /// Generate the configuration files to run a testbed of storage nodes.
+    GenerateDryRunConfigs(GenerateDryRunConfigsArgs),
```
Comment (mlegner) [inline crates/walrus-service/bin/main.rs:83]: Remark: Now that I think about this some more, I feel that those don't actually belong in the storage-node CLI. They would fit better into a separate "admin" CLI. Could be changed in a separate PR.

File: crates/walrus-service/src/testbed.rs [hunk 2]
```diff
@@ -39,146 +40,60 @@ pub fn node_config_name_prefix(node_index: u16, committee_size: NonZeroU16) -> S
...
-/// Deterministically generate storage node configurations without a `SuiConfig`.
-///
-/// The storage node configurations are created for a total of `committee_size` nodes and
-/// `n_shards`.
-pub fn create_local_storage_node_configs(
-    working_dir: &Path,
-    committee_size: NonZeroU16,
-    n_shards: NonZeroU16,
-) -> (Vec<StorageNodeConfig>, Vec<SuiStorageNode>) {
+/// Generate deterministic keypairs for the benchmark purposes.
```
Comment (mlegner) [inline crates/walrus-service/src/testbed.rs]: Nit: We generally use indicative instead of imperative mood, aligned with [RFC 505](https://github.com/rust-lang/rfcs/blob/master/text/0505-api-comment-conventions.md#formatting).

File: crates/walrus-service/src/testbed.rs [hunk 3]
```diff
@@ -39,146 +40,60 @@ pub fn node_config_name_prefix(node_index: u16, committee_size: NonZeroU16) -> S
...
+    node_index: Option<u16>,
+    committee_size: Option<u16>,
+) -> SocketAddr {
+    let port = if let (Some(node_index), Some(committee_size)) = (node_index, committee_size) {
+        port + committee_size + node_index
+    } else {
+        port
+    };
+    SocketAddr::new(IpAddr::V4(ip), port)
+}
```
Comment (mlegner) [inline crates/walrus-service/src/testbed.rs]: Remark: It is not fully clear from the documentation that this should be called with the same port as `get_metrics_address` when going for the adjustment version.
Maybe it would be better to have two separate functions instead of the version with the `Option<_>` parameters? (Same for the `get_metrics_address` above.)

Also, there is a slight simplification:
```suggestion
/// Get the REST API address for a node. If both the node index and the committee size is provided,
/// the port is adjusted to ensure uniqueness across nodes.
pub fn get_rest_api_address(
    ip: Ipv4Addr,
    mut port: u16,
    node_index: Option<u16>,
    committee_size: Option<u16>,
) -> SocketAddr {
    if let (Some(node_index), Some(committee_size)) = (node_index, committee_size) {
        port += committee_size + node_index
    };
    SocketAddr::new(IpAddr::V4(ip), port)
}
```

File: crates/walrus-service/src/testbed.rs [hunk 4]
```diff
@@ -39,146 +40,60 @@ pub fn node_config_name_prefix(node_index: u16, committee_size: NonZeroU16) -> S
...
+/// Generate deterministic and even shard allocation for the benchmark purposes.
+pub fn even_shards_allocation(
+    n_shards: NonZeroU16,
+    committee_size: NonZeroU16,
+) -> Vec<Vec<ShardIndex>> {
     let shards_per_node = n_shards.get() / committee_size.get();
     let remainder_shards = n_shards.get() % committee_size.get();
     let mut start = 0;
-    let mut sui_storage_node_configs = Vec::with_capacity(committee_size.get().into());
+    let mut shards_information = Vec::with_capacity(committee_size.get() as usize);
```
Comment (mlegner) [inline crates/walrus-service/src/testbed.rs]: Nit: I generally dislike the `as` conversions with integer types as it is unclear if this is a (potentially) lossy or lossless conversion without looking at the types.

File: crates/walrus-service/src/testbed.rs [hunk 5]
```diff
@@ -187,14 +102,22 @@ pub async fn deploy_walrus_contract(
...
-    let mut keypairs = benchmark_keypairs(committee_size);
+
+    // Check whether the testbed collocates the storage nodes on the same machine
+    // (that is, local testbed).
+    let ip_set = ips.iter().collect::<HashSet<_>>();
+    let collocated = ip_set.len() != ips.len();
+
+    // Build one Sui storage node config for each storage node.
+    let committee_size = ips.len() as u16;
+    let mut keypairs = benchmark_keypairs(committee_size as usize);
```
Comment (mlegner) [inline crates/walrus-service/src/testbed.rs:118]: Important: This way of implicitly defining the deployment mode (local vs. distributed) feels a bit dangerous to me. For example, if one accidentally has a duplicate IP address in the list, we would conclude that we are in the collocated scenario. Do we expect a scenario where some nodes have the same IP address but not all?

Suggestion: Distinguish between a local scenario with options `committee_size` and `ip_address` and a distributed one with `ip_addresses`, which are then expected and checked to be unique.

If we keep the current behavior, I would add this to the docstring for `ips` in `main.rs`.

Other remark: This code is duplicated with `create_storage_node_configs` below.

File: crates/walrus-service/src/testbed.rs (hunk 2 above)
Comment (asonnino) [reply crates/walrus-service/src/testbed.rs]: good point I forgot about that

File: crates/walrus-service/src/testbed.rs (hunk 4 above)
Comment (asonnino) [reply crates/walrus-service/src/testbed.rs]: same, I removed the conversion as it is not really needed here

File: crates/walrus-service/src/testbed.rs (hunk 3 above)
Comment (asonnino) [reply crates/walrus-service/src/testbed.rs]: Good point, I made it clearer by indicating that the function "formats" the address

File: crates/walrus-service/src/testbed.rs (hunk 5 above)
Comment (asonnino) [reply crates/walrus-service/src/testbed.rs:118]: Good point, I suggest to fix it in a different PR to not overly modify this branch (as it is currently deployed and used by Giacomo). I will make the PR immediatly

Comment (asonnino) [reply crates/walrus-service/src/testbed.rs:118]: [#399](https://github.com/MystenLabs/walrus/pull/399)

File: Cargo.lock
Comment (mlegner) [inline Cargo.lock:1]: Remark: I'm not opposed to upgrading our indirect dependencies, but I'd suggest doing this in a separate PR.

Comment (asonnino) [reply Cargo.lock:1]: I don't know what happened to cargo lock, I didn't meant to change it

Comment (asonnino) [reply Cargo.lock:1]: I can try to remove it, hopefully it won't break anything
